### PR TITLE
Linux101: environment variables

### PR DESCRIPTION
**Pull request by core-man**
Related to #67.

**References**
- https://101.lug.ustc.edu.cn/Ch06/#_3
- http://billie66.github.io/TLCL/book/chap12.html
- https://www.bookstack.cn/read/bash-tutorial/docs-variable.md
- https://wangdoc.com/bash/variable.html

### CODE REVIEW

File: source/computer/environment-variable.rst [hunk 1]
```diff
@@ -0,0 +1,61 @@
+环境变量
+========
+
+当用户登录 Linux 系统后，Shell 程序启动并读取一系列启动文件，这些文件定义了可供所有用户共享 Shell 环境。
```
Comment (seisman) [inline source/computer/environment-variable.rst:4]:
```suggestion
当用户登录 Linux 系统后，Shell 程序启动并读取一系列配置文件，这些文件设置了可供所有用户共享的 Shell 环境。
```

File: source/computer/environment-variable.rst [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+然后，Shell 会读取一些位于家目录下的启动文件，这些启动文件定义了用户个人的 Shell 环境。
```
Comment (seisman) [inline source/computer/environment-variable.rst:5]:
```suggestion
然后，Shell 会读取一些位于家目录下的配置文件，这些配置文件定义了用户个人的 Shell 环境。
```

File: source/computer/environment-variable.rst [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+这些启动文件的名字和启动顺序在在不同的 Shell 中或不同情况的 Shell 会话下，有所区别。Bash 常见的启动文件有
```
Comment (seisman) [inline source/computer/environment-variable.rst:6]:
```suggestion
这些配置文件的名字和读取顺序在不同的 Shell 中或不同情况的 Shell 会话下，有所区别。Bash 常见的配置文件有
```

File: source/computer/environment-variable.rst [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+   SHELL              Shell 程序的名字
+   HOME               当前用户的家目录
+   LD_LIBRARY_PATH    程序加载运行期间，动态链接库的查找路径
+   PATH               由冒号分开的用于搜索可执行程序名的目录列表
+   PS1                Shell 提示符
+   PWD                当前工作目录
+   USER               当前用户的用户名
+   ================== =======================================
+
+例如，\ :file:`~/.bash_profile` 或其他启动文件中可能含有以下一行::
```
Comment (seisman) [inline source/computer/environment-variable.rst:26]:
```suggestion
例如，\ :file:`~/.bashrc` 中可能含有以下一行::
```

File: source/computer/environment-variable.rst [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    export PATH=$PATH:$HOME/bin
+
+这里面出现了两个环境变量，即\ ``HOME`` 和 ``PATH``\ ：
+
+- ``HOME`` 表示用户的家目录，如 :file:`/home/seismo-learn`
+- ``PATH`` 表示由冒号分开的目录列表。当输入可执行程序名后，Shell 会搜索这个目录列表来寻找可执行程序。
+
+变量前加美元符号 ``$`` 表示读取变量值。因此，这一行的意思是将 :file:`/home/seismo-learn/bin` 目录
+加入可执行程序的搜索目录列表。这也是之前示例中我们直接在终端输入命令后，Shell 能正确找到命令所在位置的原因。
+若某一可执行程序所在目录不在 ``PATH`` 中，则 Shell 无法将其找到交给系统执行。
```
Comment (seisman) [inline source/computer/environment-variable.rst:37]:
```suggestion
若某一可执行程序所在目录不在 ``PATH`` 所包含的目录中中，则 Shell 无法将其找到交给系统执行。
```

File: source/computer/environment-variable.rst [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+可以使用 ``which`` 或 ``whereis`` 命令查询命令所在目录::
+
+    $ which cd
+    /usr/bin/cd
+    $ whereis cd
+    cd: /usr/bin/cd /usr/share/man/man1/cd.1.gz /usr/share/man/man1p/cd.1p.gz /usr/share/man/mann/cd.n.gz
```
Comment (seisman) [inline source/computer/environment-variable.rst:44]: On macOS + zsh:
```
which cd
cd: shell built-in command
```

Here, `cd` is a shell built-in command, so `which cd` doesn't work. Need to change to another command.

File: source/computer/environment-variable.rst [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+可以使用以下命令查看环境变量::
+
+    # 使用 printenv 命令查看 SHELL 变量，即 Shell 程序的名字
+    $ printenv SHELL
+    /bin/bash
```
Comment (seisman) [inline source/computer/environment-variable.rst:50]: `echo ${SHELL}` is enough.
```suggestion
```

File: source/computer/environment-variable.rst [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+    # 也可以使用 echo 命令查看
+    $ echo ${SHELL}
+
+    # 查看所有环境变量，并使用管道和 sort 命令将他们按照字符顺序排列
+    $ printenv | sort
+    SHELL=/bin/bash
+    USER=seismo-learn
+    # 也可以使用 set 命令，不过 Shell 变量 和 Shell 函数也同时会被显示
+    $ set
```
Comment (seisman) [inline source/computer/environment-variable.rst:59]: Useless for newbies, right?
```suggestion
```

File: source/computer/environment-variable.rst [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+
+日常科研中，我们一般修改 :file:`~/.bashrc` 文件来修改环境变量以配置个人的 Shell 环境。
```
Comment (seisman) [inline source/computer/environment-variable.rst:61]: Perhaps we should mention `~/.zshrc` for macOS and zsh users.

File: source/computer/environment-variable.rst [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+    /usr/bin/cd
+    $ whereis cd
+    cd: /usr/bin/cd /usr/share/man/man1/cd.1.gz /usr/share/man/man1p/cd.1p.gz /usr/share/man/mann/cd.n.gz
+
+可以使用以下命令查看环境变量::
+
+    # 使用 printenv 命令查看 SHELL 变量，即 Shell 程序的名字
+    $ printenv SHELL
+    /bin/bash
+    # 也可以使用 echo 命令查看
```
Comment (core-man) [inline source/computer/environment-variable.rst:51]:
```suggestion
    # 使用 echo 命令查看 SHELL 变量，即 Shell 程序的名字
```

File: source/computer/environment-variable.rst (hunk 9 above)
Comment (core-man) [reply source/computer/environment-variable.rst]: Added in 2f2e9c2.

File: source/computer/environment-variable.rst (hunk 6 above)
Comment (core-man) [reply source/computer/environment-variable.rst]: Fixed in 2f2e9c2.

File: source/computer/environment-variable.rst [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+   PS1                Shell 提示符
+   PWD                当前工作目录
+   USER               当前用户的用户名
+   ================== =======================================
+
+例如，\ :file:`~/.bashrc` 中可能含有以下一行::
+
+    export PATH=$PATH:$HOME/bin
+
+这里面出现了两个环境变量，即\ ``HOME`` 和 ``PATH``\ ：
```
Comment (seisman) [inline source/computer/environment-variable.rst:30]: Explain why `export` is necessary.

Comment (core-man) [reply source/computer/environment-variable.rst]: Done in 59ed818.

File: source/computer/environment-variable.rst [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+   SHELL              Shell 程序的名字
+   HOME               当前用户的家目录
+   LD_LIBRARY_PATH    程序加载运行期间，动态链接库的查找路径
+   PATH               由冒号分开的用于搜索可执行程序名的目录列表
+   PS1                Shell 提示符
+   PWD                当前工作目录
+   USER               当前用户的用户名
+   ================== =======================================
+
+使用  echo 命令查看环境变量::
```
Comment (seisman) [inline source/computer/environment-variable.rst:24]:
```suggestion
使用  ``echo`` 命令查看环境变量::
```

File: source/computer/environment-variable.rst [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+    $ echo ${SHELL}
+
+使用 ``which`` 或 ``whereis`` 命令可查询命令所在目录::
+
+    $ which cat
+    /usr/bin/cat
+
+    $ whereis cat
+    cat: /usr/bin/cat /usr/share/man/man1/cat.1.gz /usr/share/man/man1p/cat.1p.gz
+
```
Comment (seisman) [inline source/computer/environment-variable.rst:36]: It's unclear to me why we want to mention `which` and `whereis` here.

File: source/computer/environment-variable.rst [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+这里使用了两个环境变量（\ ``HOME`` 和 ``PATH``\ ）和一个 ``export`` 命令：
+
+- ``HOME`` 表示用户的家目录，如 :file:`/home/seismo-learn`
+- ``PATH`` 表示由冒号分开的目录列表。当输入可执行程序名后，Shell 会搜索这个目录列表来寻找可执行程序。
+- ``export`` 命令用于新增、修改或删除环境变量
+
+变量前加美元符号 ``$`` 表示读取变量值。因此，这一行的意思是将 :file:`/home/seismo-learn/bin` 目录
+加入可执行程序的搜索目录列表。这也是之前示例中我们直接在终端输入命令后，Shell 能正确找到命令所在位置的原因。
+若某一可执行程序所在目录不在 ``PATH`` 所包含的目录中，则 Shell 无法将其找到交给系统执行。
```
Comment (seisman) [inline source/computer/environment-variable.rst:50]: This paragraph is still unclear. Perhaps we should `echo $PATH` first, and then run `export PATH=$PATH:$HOME/bin`, and show `echo $PATH` again?

File: source/computer/environment-variable.rst (hunk 13 above)
Comment (core-man) [reply source/computer/environment-variable.rst]: I'd like to introduce those two commands in Linux101. Now I hid them in 49c6363.

File: source/computer/environment-variable.rst (hunk 14 above)
Comment (core-man) [reply source/computer/environment-variable.rst]: Done in 49c6363.

**Comment by core-man** [comment]
I haved improved it.

**Review by seisman — APPROVED** [review]

**Pull request merged**